### PR TITLE
SNOW-808908: Handle Reserved Keywords and Special Characters in Schematization

### PR DESCRIPTION
**Pull request by sfc-gh-tzhang**
To Handle Reserved Keywords and Special Characters in Schematization, we will convert all names to uppercase and then add double quotes, except for the names that have double quotes already. By doing this, we tell compiler to bypass the reserved keyword or special characters check, and we’re also preserving the old behavior for customers
Examples:

- start -> “START”
- a@b -> “A@B”
- c1/C1 -> c1, C1, “C1” would all work
- “ab” -> “ab”

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/Utils.java [hunk 1]
```diff
@@ -137,6 +138,94 @@ public class Utils {
 
   private static final KCLogger LOGGER = new KCLogger(Utils.class.getName());
 
+  private static final Pattern unquotedIdentifierPattern =
+      Pattern.compile("[$][0-9]+|[_a-zA-Z][_a-zA-Z0-9]*[$]?[_a-zA-Z0-9]*");
+
+  private static final String[] reservedKeywords = {
```
Comment (sfc-gh-xhuang) [inline src/main/java/com/snowflake/kafka/connector/Utils.java:144]: Do we need to put these reservedKeywords here? It will not always be kept in sync / be the full ist

**Comment by sfc-gh-tzhang** [comment]
ping, can I get a review on this one?

File: src/main/java/com/snowflake/kafka/connector/Utils.java [hunk 2]
```diff
@@ -137,6 +137,18 @@ public class Utils {
 
   private static final KCLogger LOGGER = new KCLogger(Utils.class.getName());
 
+  /**
+   * Quote the column name if needed: When there is quote already, we do nothing; otherwise we
+   * convert the name to upper case and add quote
+   */
+  public static String quoteNameIfNeeded(String name) {
+    int length = name.length();
+    if (name.charAt(0) == '"' && (length >= 2 && name.charAt(length - 1) == '"')) {
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/Utils.java:146]: can use startswith and endswith?

File: src/test/java/com/snowflake/kafka/connector/internal/TestUtils.java [hunk 3]
```diff
@@ -354,8 +355,16 @@ static Map<String, String> getConfWithEncryptedKey() {
   }
 
   public static Map<String, String> getConfWithOAuth() {
+    // Return null if we don't have the corresponding OAuth required configuration in the config
+    // file
+    Map<String, String> conf = getConf();
```
Comment (sfc-gh-japatel) [inline src/test/java/com/snowflake/kafka/connector/internal/TestUtils.java:360]: can you make oauth changes in separate PRs?

File: src/test/java/com/snowflake/kafka/connector/records/RecordContentTest.java [hunk 4]
```diff
@@ -271,6 +271,6 @@ public void testColumnNameFormatting() throws JsonProcessingException {
     Map<String, Object> got = service.getProcessedRecordForStreamingIngest(record);
 
     assert got.containsKey("\"NaMe\"");
-    assert got.containsKey("AnSwEr");
+    assert got.containsKey("\"ANSWER\"");
```
Comment (sfc-gh-japatel) [inline src/test/java/com/snowflake/kafka/connector/records/RecordContentTest.java:274]: to understand again, is this because name is double quoted but answer is not?

**Review by sfc-gh-japatel — COMMENTED** [review]
does any of the test we have connects to JDBC with this column name changes and verifies the behavior?
I also think this deserves a new End to end test. WDYT

Overall lgtm! Thank you Toby!

File: src/main/java/com/snowflake/kafka/connector/Utils.java [hunk 5]
```diff
@@ -137,6 +137,18 @@ public class Utils {
 
   private static final KCLogger LOGGER = new KCLogger(Utils.class.getName());
 
+  /**
+   * Quote the column name if needed: When there is quote already, we do nothing; otherwise we
+   * convert the name to upper case and add quote
+   */
+  public static String quoteNameIfNeeded(String name) {
```
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/Utils.java:144]: add unit tests for this method

**Review by sfc-gh-rcheng — COMMENTED** [review]
lgtm, need unit tests for the quoteNameIfNeeded method

File: src/main/java/com/snowflake/kafka/connector/Utils.java (hunk 2 above)
Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/Utils.java:146]: No strong preference here, but charAt should performance better than startswith? I see startswith is usually used for matching a string instead of a character

File: src/main/java/com/snowflake/kafka/connector/Utils.java (hunk 5 above)
Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/Utils.java:144]: added

File: src/test/java/com/snowflake/kafka/connector/records/RecordContentTest.java (hunk 4 above)
Comment (sfc-gh-tzhang) [reply src/test/java/com/snowflake/kafka/connector/records/RecordContentTest.java:274]: correct

**Comment by sfc-gh-tzhang** [comment]
> does any of the test we have connects to JDBC with this column name changes and verifies the behavior? I also think this deserves a new End to end test. WDYT
> 
> Overall lgtm! Thank you Toby!

We many both IT and end2end testing for this. For IT, one example is testSchematizationWithTableCreationAndAvroInput where we insert the rows and compare the result by scanning the table. For end2end, one example is TestSchemaEvolutionJson where we test names with and without double quotes, and compare it with gold_type in the end. Let me know if you think anything is missing here.

**Review by sfc-gh-rcheng — APPROVED** [review]
